### ISSUE TITLE
Base58

### ISSUE THREAD

**Issue by mik3hall** [blockchain/api-v1-client-java#12]
To make a valid issue, although I have additional questions. This might be considered a enhancement request to include Base58 conversions as part of the API.

[end of enhancement request]
_____

Additional questions. (Please feel free to indicate if there is a more appropriate place to ask these or where there is documentation I should refer to before asking).

I am new to this and just got into where the code I'm working on started looking at real transactions. I am interested in the crypto parts of these. The address and signatures, hashes I am assuming are strictly hex encoded byte arrays.  
Right now I'm having trouble getting pretty much any of my values to match up. Looking at it it seems encoding might part of the difficulty. 
Do both addresses and signatures use Base58 encoding, or the bitcoin variant?
I am now assuming at least addresses do. I am looking for now at using the bitcoinj Base58 class although introducing a second api probably with different licensing isn't an ideal situation. I may attempt my own at some point if such a class doesn't seem like a fit to the blockchain.info api's.
To further make sure I am understanding the api correctly. Given a output script like...

[java] *** Script ***
     [java] OP_DUP
     [java] OP_HASH160
     [java] 396652927daedcef95e1fb89a09faf09545c3225
     [java] OP_EQUALVERIFY
     [java] OP_CHECKSIG
     [java] ***************

Coming into it I believe signature and public key/address are expected to be on the stack.
For the api would these come from...

signature - input.getScriptSignature(); would that be Base58 or some other encoding?

address - Output prior = input.getPreviousOutput();
	                 System.out.println("prior output is " + prior);
	                 String addr = prior.getAddress();
This I'm currently assuming is Base58 and is the pubic key expected to be on the stack
.
Would this be correct api java to get <sig> and <pubkey> on the stack before running the output script?

Thanks

**Comment by matsjj** [comment]
Indeed a B58 encoder is necessary in many instances. I added one to etc.Base58. It is in fact derived from the bitcoinj project, but since they are both MIT license, this is not an issue. Also some added convenience to not add an additional project.

I am happy to help with your additional problems, but I have troubles understanding them. Do you mind elaborating what you are trying to do and where it fails? B58 is not used internally in bitcoin (at all!), it is just used whenever some of the data touches a UI (like displaying an address). If you want to validate the script, you need the pubkey and signature on the stack, both in hex format and little endian notation.

**Comment by mik3hall** [comment]
> On Oct 13, 2015, at 5:17 AM, matsjj <notifications@github.com> wrote:
> 
> Indeed a B58 encoder is necessary in many instances. I added one to etc.Base58. It is in fact derived from the bitcoinj project, but since they are both MIT license, this is not an issue. Also some added convenience to not add an additional project.
> 
> I am happy to help with your additional problems, but I have troubles understanding them. Do you mind elaborating what you are trying to do and where it fails? B58 is not used internally in bitcoin (at all!), it is just used whenever some of the data touches a UI (like displaying an address). If you want to validate the script, you need the pubkey and signature on the stack, both in hex format and little endian notation.
> 
bitcoinj I thought was an apache license but no matter at this point. 

Background
I am taking a online MOOC course, Coursera - Bitcoin and Cryptocurrency Technologies <https://www.coursera.org/course/bitcointech>. My original idea was just a command line tool that could be used by someone to familiarize themselves with some of the bitcoin related content like scripting and maybe get some hands on with the crypto. My current thought is to actually to give it some useful functionality as a command line tool. Not sure exactly what yet, hopefully not just duplicating things blockchain.info or other code does. Maybe simulating data and identities and being able to develop your own scripts and test it on that data, for one idea.

Specific to issue
For much of this I did think I would just use the blockchain.info api’s. To start out, for working with actual ‘live’ blockchain data. I am trying to figure out if I can take existing transactions and run the simple verification script provided in the original issue. Specific to the api I am trying to figure out what api methods correspond to fetching the required public key and signature parameters and how these fields are encoded. The problem I was running into at this point is that nothing so far is matching as far as hashing or signatures. When i realized address was probably Base58 encoded I sort of stopped worrying about that, until I have time to look at that. It does seem to me if anyone using the java api’s wants to do anything crypto related with the blockchain data they are going to have deal with the base58 address encoding?  
So basically all I’m looking for is a simple yes/no on whether I am looking at the correct api methods for public key and signature. Also, some api provided way to deal with base58 as per the enhancement request, which you say is now there. Possibly one more thing would be indicating in the api documentation indicating a fields encoding if not obvious. The base58 sort of caught me by surprise and took me a while to figure out. Again, I am learning most of this myself as I go along.

Thanks

Michael Hall

**Comment by matsjj** [comment]
> bitcoinj I thought was an apache license but no matter at this point.

Hm might be, but the Base58 file was under MIT at least...

That course is very good, good decision to work your way through it! 

The scripting system is kinda difficult, so another tool would be great to help other people familiarise with it. You want to explore how many wheels you want to redefine though. For example bitcoinj does have a great Script evaluation built in already (see Script.correctlySpends(...) ).

You can obtain the output script of a transaction directly via our API (transaction.outputs.0..9.script). It is a byte[], just as it is stored in the blockchain. This means that you don't necessarily need a B58 encoder, as the address is just a representation of a special (or rather common) form of script. 

So for the transaction
https://blockchain.info/tx/6d50161414558cb71ba7886be0c1c300e2b18667509a983ebfac819b0c546d81
for example, you have two outputs, and the second one is
> 76a914c5b7fd920dce5f61934e792c7e6fcc829aff533d88ac
which basically translate into 
> 'pay to the pubkeyhash '14c5b7fd920dce5f61934e792c7e6fcc829aff533d88''
which we translate to the address 
> 1K2SXgApmo9uZoyahvsbSanpVWbzZWVVMF

You can check the input for this transaction https://blockchain.info/tx-index/106800549
and also via our API.

**Comment by mik3hall** [comment]
> 
> On Oct 16, 2015, at 3:42 AM, matsjj <notifications@github.com> wrote:
> 

> For example bitcoinj does have a great Script evaluation built in already (see Script.correctlySpends(...) ).
> 

I’m hoping my code will give a better, more interactive, feeling for how the scripts work.


> the address is just a representation of a special (or rather common) form of script.

I think sometimes this is true but sometimes the address is the Base58 encoded public key?

I found a pretty good reference I can use on this and will see if I can work through the issues I was having with that.

Thanks.

Michael Hall

**Comment by matsjj** [comment]
> I’m hoping my code will give a better, more interactive, feeling for how the scripts work.

Oh yes, bitcoinj is not supposed to give any interactive feeling about that. ;)
I'm saying that you should probably not tinker with actual script evaluation, but rather with how to get the point of proper scripts across to new developers. 

>I think sometimes this is true but sometimes the address is the Base58 encoded public key?
>I found a pretty good reference I can use on this and will see if I can work through the issues I was having with that.

I'm not 100% sure on that, especially since I haven't looked up an example, but there are 4 different transactions you can have
Pay 2 PubKey
Pay 2 PubKeyHash
Pay 2 ScriptHash
OP_RETURN
where only P2PKH and P2SH are representable as addresses. When you have P2PK or OP_RETURN, they are in general not displayed as addresses, although you could derive a P2PKH address from a P2PK transaction.

**Comment by mik3hall** [comment]
To the issue. I tried the Base58 code you added to etc. and it worked fine. You could probably close this if you like.

**Comment by mik3hall** [comment]
To your last. Trying to get my code to run real blockchains transactions I think will be useful for someone trying to understand scripting. They can choose a transaction, run it's script, see it works and then go, "well yes I see this does work with real data." 
For me, I think it's very important to my code to make sure this works. How can I be sure I'm telling anyone anything useful about scripts if I can't say I can myself successfully handle those scripts?
Made a little progress this week, I can get through the OP_HASH160 in a 'Pay 2 PubKeyHash'. Quite a ways to go though.

**Comment by matsjj** [comment]
Yes, it is definitely a learning experience! And if you want to go through at least some of the OP_codes, it will be great for you to implement those.
